extensions/socket/internal.m

```diff
@@ -0,0 +1,362 @@
+/// Method
+/// Write data to the socket
+///
+/// Parameters:
+///  * message - A string containing data to be sent on the socket
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_write(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TSTRING, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+    NSString *message = [skin toNSObjectAtIndex:2];
+
+    [asyncSocket writeData:[message dataUsingEncoding:NSUTF8StringEncoding ] withTimeout:-1 tag:-1];
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket:setCallback([fn]) -> self
+/// Method
+/// Sets the callback for the socket. Required for working with read data
+///
+/// Parameters:
+///  * fn - An optional callback function with single parameter containing data read from the socket. A `nil` argument or nothing clears the callback, defaulting to `print`
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_setCallback(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TFUNCTION|LS_TNIL|LS_TOPTIONAL, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+
+    if (lua_type(L, 2) == LUA_TFUNCTION) {
+        lua_pushvalue(L, 2);
+        asyncSocket.callback = [skin luaRef:refTable];
+    } else {
+        asyncSocket.callback = 0;
+    }
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket:connect() -> self
+/// Method
+/// Connects an unconnected `hs.socket` instance
+///
+/// Parameters:
+///  * host - A string containing the hostname or IP address
+///  * port - A port number [1024-65535]. Ports [1-1023] are privileged
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_connect(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TSTRING, LS_TNUMBER, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+
+    NSString *theHost = [skin toNSObjectAtIndex:2];
+    NSNumber *thePort = [skin toNSObjectAtIndex:3];
+
+    NSError *err;
+    if (![asyncSocket connectToHost:theHost onPort:[thePort unsignedShortValue] error:&err]) {
+        [skin logError:[NSString stringWithFormat:@"Unable to connect: %@", err]];
+    }
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket:disconnect() -> self
+/// Method
+/// Disconnects the socket instance
+///
+/// Parameters:
+///  * None
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_disconnect(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+    [asyncSocket disconnect];
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
```

Comment:
I'd suggest we have an hs.socket:connected() here, even though the same information is in :info()